### PR TITLE
initial proposal for feature in #330

### PR DESCRIPTION
**Pull request by runar-rkmedia**
## What

See #330 for specifics, here is a short recap:

> When using breakpoints, it may be desirable to only receive updates if the breakpoint changes, and not on tiny changes to width/height etc.

## How

Two new options are added:

```typescript
interface SetState {
  (previous: State, next: State): State;
}
export interface Options<T> {
  ///...ommited for brevity
  onlyUpdateOnBreakpointChange: boolean
  customSetState: SetState
}
```

## Checklist

Have you done all of these things?

- [x] Documentation added
- [X] Tests
- [X] TypeScript definitions updated
- [X] Ready to be merged

### CODE REVIEW

**Comment by wellyshen** [comment]
@runar-rkmedia Thank you for this PR, there're some ideas I'd like to discuss with you.

File: src/index.ts [hunk 1]
```diff
@@ -62,17 +72,23 @@ const useDimensions = <T extends HTMLElement>({
   });
   const prevSizeRef = useRef<{ width?: number; height?: number }>({});
   const prevBreakpointRef = useRef<string>();
-  const observerRef = useRef<ResizeObserver | null>(null);
+  const observerRef = useRef<any | null>(null);
```
Comment (wellyshen) [inline src/index.ts:75]: @runar-rkmedia Why we need to change this type?

**Comment by runar-indico** [comment]
Sure thing, @wellyshen Would you like to discuss this here or some other platform?

File: src/index.ts [hunk 2]
```diff
@@ -19,12 +19,20 @@ interface Event<T> extends State {
...
 export interface Options<T> {
   ref?: RefObject<T>;
   useBorderBoxSize?: boolean;
   breakpoints?: Breakpoints;
   onResize?: OnResize<T>;
+  /** If set, it will only update the state when a breakpoint is changed. */
+  onlyUpdateOnBreakpointChange?: boolean;
+  /** If you wish to conditionally update the internal state, for instance to
+   * reduce rerenders conditionally, you may use this custom-functionality */
+  customSetState?: SetState;
```
Comment (wellyshen) [inline src/index.ts:35]: @runar-rkmedia How about the following idea?

```js
const {/* ... */} = useDimensions({
  shouldUpdate: (prevState, nextState) => prevState.width > 300,
});
```

File: src/index.ts [hunk 3]
```diff
@@ -145,12 +161,28 @@ const useDimensions = <T extends HTMLElement>({
...
-      });
+      };
+      if (customSetStateRef.current) {
+        setState((s) =>
+          customSetStateRef.current ? customSetStateRef.current(s, next) : next
+        );
+        return;
+      }
+      if (breakpoints && onlyUpdateOnBreakpointChange) {
+        setState((s) => {
```
Comment (wellyshen) [inline src/index.ts:177]: @runar-rkmedia We can simplify the code as below:

```js
if (breakpoints && onlyUpdateOnBreakpointChange ) {
  setState((s) => s.currentBreakpoint === e.currentBreakpoint ? s : next);
  return;
}
```

**Review by wellyshen — COMMENTED** [review]
@runar-rkmedia Thank you for your effort, let's make this library better via the discussion :)

**Comment by wellyshen** [comment]
@runar-indico Wow you have two accounts, I think we can discuss here via the review.

File: src/index.ts (hunk 1 above)
Comment (runar-indico) [reply src/index.ts:75]: sorry, guess I missed that one during developmen. I'll revert that change.

File: src/index.ts (hunk 2 above)
Comment (runar-indico) [reply src/index.ts:35]: Yes, I agree, the naming here is a lot better.

**Comment by runar-rkmedia** [comment]
Yeah, sorry for using two accounts here. I switched to my personal account again.

File: src/index.ts (hunk 3 above)
Comment (runar-rkmedia) [reply src/index.ts:177]: I'll update with your suggestions here.

File: src/index.ts (hunk 2 above)
Comment (wellyshen) [reply src/index.ts:35]: @runar-rkmedia I also think if we can return a `boolean` to tell the state to update or not, like the `shouldComponentUpdate` lifecycle of React.

**Comment by wellyshen** [comment]
@runar-rkmedia I think the code looks great! But I'm not sure if you're ready, plz feel free to let me know that.

**Comment by runar-rkmedia** [comment]
I think it is ready.

Thank you for your quick feedback here, I really appreciate it.

**Comment by wellyshen** [comment]
@runar-rkmedia Nice! Thank you for let this library getting better. I will release the features soon once I finish the docs and unit testing.

@all-contributors please add @runar-rkmedia  for code, doc, idea.

**Pull request merged**